File: src/core/facade-extractor.ts
// ============================================================================
// Facade Extractor
//
// Takes Face3D[] and produces Facade[] — one elevation view per building side.
//
// Algorithm:
//   1. Auto-detect vertical axis (Y-up vs Z-up).
//   2. Filter vertical faces.
//   3. Cluster faces by horizontal direction → N/S/E/W groups.
//   4. For each group, collect unique outline edges (removing shared internal
//      edges from triangulated meshes) to produce clean silhouettes.
//   5. Normalize coordinates so (0,0) is bottom-left.
// ============================================================================

import type { Face3D, Facade, Loop2D, Vec2, Vec3 } from "./types";
import { cross, dot, normalize, vlength } from "./types";

const VERTICAL_EPSILON = 0.20;
const DIRECTION_CLUSTER_THRESHOLD = 0.70;

type UpAxis = "Y" | "Z";

function getUpComponent(normal: Vec3, up: UpAxis): number {
  return up === "Y" ? normal.y : normal.z;
}

function getUpVec(up: UpAxis): Vec3 {
  return up === "Y" ? { x: 0, y: 1, z: 0 } : { x: 0, y: 0, z: 1 };
}

function horizontalDir(normal: Vec3, up: UpAxis): Vec3 {
  const h: Vec3 =
    up === "Y"
      ? { x: normal.x, y: 0, z: normal.z }
      : { x: normal.x, y: normal.y, z: 0 };
  return normalize(h);
}

function computeFacadeAxes(
  direction: Vec3,
  up: UpAxis,
): { uAxis: Vec3; vAxis: Vec3 } {
  const worldUp = getUpVec(up);
  const uAxis = normalize(cross(worldUp, direction));
  const vAxis = worldUp;
  return { uAxis, vAxis };
}

function clusterByDirection(
  faces: Face3D[],
  up: UpAxis,
): Array<{ dir: Vec3; faces: Face3D[] }> {
  const clusters: Array<{ dir: Vec3; faces: Face3D[] }> = [];

  for (const face of faces) {
    const hDir = horizontalDir(face.normal, up);
    if (vlength(hDir) < 0.01) continue;

    let placed = false;
    for (const cluster of clusters) {
      if (dot(hDir, cluster.dir) > DIRECTION_CLUSTER_THRESHOLD) {
        cluster.faces.push(face);
        placed = true;
        break;
      }
    }

    if (!placed) {
      clusters.push({ dir: hDir, faces: [face] });
    }
  }

  return clusters;
}

function directionLabel(direction: Vec3, up: UpAxis): string {
  const angle =
    up === "Y"
      ? Math.atan2(direction.x, direction.z)
      : Math.atan2(direction.x, direction.y);
  const deg = ((Math.atan2(Math.sin(angle), Math.cos(angle)) * 180) / Math.PI + 360) % 360;

  if (deg < 45 || deg >= 315) return "Fachada Norte";
  if (deg < 135) return "Fachada Este";
  if (deg < 225) return "Fachada Sur";
  return "Fachada Oeste";
}

/** Round a coordinate to a fixed precision for use as an edge key. */
function roundCoord(v: number): number {
  return Math.round(v * 10000) / 10000;
}

function edgeKey(ax: number, ay: number, bx: number, by: number): string {
  const a = `${roundCoord(ax)},${roundCoord(ay)}`;
  const b = `${roundCoord(bx)},${roundCoord(by)}`;
  return a < b ? `${a}|${b}` : `${b}|${a}`;
}

function extractWithAxis(faces: Face3D[], up: UpAxis): Facade[] {
  // 1. Filter vertical faces.
  const verticalFaces: Face3D[] = [];
  for (const face of faces) {
    if (Math.abs(getUpComponent(face.normal, up)) <= VERTICAL_EPSILON) {
      verticalFaces.push(face);
    }
  }
  if (verticalFaces.length === 0) return [];

  // 2. Cluster by horizontal direction.
  const clusters = clusterByDirection(verticalFaces, up);

  // 3. Build one Facade per cluster.
  const facades: Facade[] = [];

  for (const { dir, faces: clusterFaces } of clusters) {
    const { uAxis, vAxis } = computeFacadeAxes(dir, up);

    // Project all faces, collecting unique edges to remove internal
    // triangulation lines (edges shared between two faces cancel out).
    const edgeCounts = new Map<string, { ax: number; ay: number; bx: number; by: number }>();

    for (const face of clusterFaces) {
      const pts: Vec2[] = face.vertices.map((v) => ({
        x: dot(v, uAxis),
        y: dot(v, vAxis),
      }));

      // Walk edges of this polygon.
      for (let i = 0; i < pts.length; i++) {
        const j = (i + 1) % pts.length;
        const key = edgeKey(pts[i].x, pts[i].y, pts[j].x, pts[j].y);
        if (edgeCounts.has(key)) {
          edgeCounts.delete(key); // shared edge → remove
        } else {
          edgeCounts.set(key, {
            ax: pts[i].x,
            ay: pts[i].y,
            bx: pts[j].x,
            by: pts[j].y,
          });
        }
      }
    }

    if (edgeCounts.size === 0) continue;

    // Collect all points for bounding box.
    let minU = Infinity, maxU = -Infinity;
    let minV = Infinity, maxV = -Infinity;
    const edges: Array<{ ax: number; ay: number; bx: number; by: number }> = [];

    for (const edge of edgeCounts.values()) {
      edges.push(edge);
      minU = Math.min(minU, edge.ax, edge.bx);
      maxU = Math.max(maxU, edge.ax, edge.bx);
      minV = Math.min(minV, edge.ay, edge.by);
      maxV = Math.max(maxV, edge.ay, edge.by);
    }

    const width = maxU - minU;
    const height = maxV - minV;
    if (width < 0.01 || height < 0.01) continue;

    // Normalize to (0,0) origin and build Loop2D segments as 2-vertex lines.
    const polygons: Loop2D[] = edges.map((e) => ({
      vertices: [
        { x: e.ax - minU, y: e.ay - minV },
        { x: e.bx - minU, y: e.by - minV },
      ],
    }));

    let label = directionLabel(dir, up);
    const existing = new Set(facades.map((f) => f.label));
    if (existing.has(label)) {
      let n = 2;
      while (existing.has(`${label} ${n}`)) n++;
      label = `${label} ${n}`;
    }

    facades.push({ label, direction: dir, polygons, width, height });
  }

  const order: Record<string, number> = { Norte: 0, Este: 1, Sur: 2, Oeste: 3 };
  facades.sort(
    (a, b) =>
      (order[a.label.split(" ").pop()!] ?? 99) -
      (order[b.label.split(" ").pop()!] ?? 99),
  );

  return facades;
}

/** Detect the model's up axis by comparing facade extraction results for Y vs Z. */
export function detectUpAxis(faces: Face3D[]): "Y" | "Z" {
  if (faces.length === 0) return "Z";

  const facadesZ = extractWithAxis(faces, "Z");
  const facadesY = extractWithAxis(faces, "Y");

  const countZ = facadesZ.reduce((s, f) => s + f.polygons.length, 0);
  const countY = facadesY.reduce((s, f) => s + f.polygons.length, 0);

  return countY > countZ ? "Y" : "Z";
}

export function extractFacades(faces: Face3D[], upAxis?: "Y" | "Z"): Facade[] {
  if (faces.length === 0) return [];

  const up = upAxis ?? detectUpAxis(faces);
  return extractWithAxis(faces, up);
}

File: backend-selfhosted/app/core/facade_extractor.py
"""
Facade Extractor

Takes Face3D[] and produces Facade[] — one elevation view per building side.

Algorithm:
  1. Auto-detect the vertical axis (Y-up vs Z-up).
  2. Filter faces whose normal is roughly perpendicular to up (vertical faces).
  3. Compute each face's horizontal direction (the normal projected onto the
     ground plane).
  4. Cluster faces by horizontal direction into facade groups.
     Typical rectangular buildings yield 4 groups (N/S/E/W).
  5. For each group, project ALL face vertices onto the facade plane to
     produce a complete 2D elevation view with all geometry in position.
  6. Normalize coordinates so (0,0) is the bottom-left of each facade.
"""

from __future__ import annotations

import math
from typing import Literal

from .types import (
    Face3D,
    Facade,
    Loop2D,
    Vec2,
    Vec3,
    cross,
    dot,
    length,
    normalize,
    scale,
    sub,
)

VERTICAL_EPSILON = 0.20
DIRECTION_CLUSTER_THRESHOLD = 0.70  # dot product; ~45 degrees


def _get_up_component(normal: Vec3, up_axis: Literal["Y", "Z"]) -> float:
    return normal.y if up_axis == "Y" else normal.z


def _get_up_vec(up_axis: Literal["Y", "Z"]) -> Vec3:
    return Vec3(0.0, 1.0, 0.0) if up_axis == "Y" else Vec3(0.0, 0.0, 1.0)


def _horizontal_dir(normal: Vec3, up_axis: Literal["Y", "Z"]) -> Vec3:
    """Project the normal onto the ground plane and normalize."""
    if up_axis == "Y":
        h = Vec3(normal.x, 0.0, normal.z)
    else:
        h = Vec3(normal.x, normal.y, 0.0)
    return normalize(h)


def _compute_facade_axes(
    direction: Vec3, up_axis: Literal["Y", "Z"]
) -> tuple[Vec3, Vec3]:
    """Compute the facade's local 2D axes.

    Returns (u_axis, v_axis):
      u_axis = horizontal along the facade (left to right when looking at it)
      v_axis = vertical (world up)
    """
    world_up = _get_up_vec(up_axis)
    # u_axis is perpendicular to both the direction and up
    u_axis = normalize(cross(world_up, direction))
    # v_axis is just world up
    v_axis = world_up
    return u_axis, v_axis


def _compute_model_diagonal(faces: list[Face3D]) -> float:
    if not faces:
        return 1.0
    min_x = min_y = min_z = float("inf")
    max_x = max_y = max_z = float("-inf")
    for face in faces:
        for v in face.vertices:
            min_x = min(min_x, v.x)
            min_y = min(min_y, v.y)
            min_z = min(min_z, v.z)
            max_x = max(max_x, v.x)
            max_y = max(max_y, v.y)
            max_z = max(max_z, v.z)
    dx = max_x - min_x
    dy = max_y - min_y
    dz = max_z - min_z
    return math.sqrt(dx * dx + dy * dy + dz * dz)


def _cluster_by_direction(
    faces: list[Face3D], up_axis: Literal["Y", "Z"]
) -> list[tuple[Vec3, list[Face3D]]]:
    """Cluster vertical faces by their horizontal facing direction.

    Returns list of (representative_direction, faces_in_cluster).
    """
    clusters: list[tuple[Vec3, list[Face3D]]] = []

    for face in faces:
        h_dir = _horizontal_dir(face.normal, up_axis)
        if length(h_dir) < 0.01:
            continue

        placed = False
        for i, (c_dir, c_faces) in enumerate(clusters):
            dp = dot(h_dir, c_dir)
            if dp > DIRECTION_CLUSTER_THRESHOLD:
                c_faces.append(face)
                placed = True
                break

        if not placed:
            clusters.append((h_dir, [face]))

    return clusters


def _direction_label(direction: Vec3, up_axis: Literal["Y", "Z"]) -> str:
    """Give a human-readable label based on the direction."""
    if up_axis == "Y":
        # Ground plane is XZ.  +Z = "Norte", -Z = "Sur", +X = "Este", -X = "Oeste"
        angle = math.atan2(direction.x, direction.z)
    else:
        # Ground plane is XY.  +Y = "Norte", -Y = "Sur", +X = "Este", -X = "Oeste"
        angle = math.atan2(direction.x, direction.y)

    # Normalize to degrees.
    deg = math.degrees(angle) % 360

    if deg < 45 or deg >= 315:
        return "Fachada Norte"
    elif 45 <= deg < 135:
        return "Fachada Este"
    elif 135 <= deg < 225:
        return "Fachada Sur"
    else:
        return "Fachada Oeste"


def _face_center(face: Face3D) -> Vec3:
    """Compute the centroid of a face."""
    n = len(face.vertices)
    if n == 0:
        return Vec3(0.0, 0.0, 0.0)
    sx = sum(v.x for v in face.vertices)
    sy = sum(v.y for v in face.vertices)
    sz = sum(v.z for v in face.vertices)
    return Vec3(sx / n, sy / n, sz / n)


def _filter_exterior_faces(
    cluster_faces: list[Face3D], direction: Vec3, model_diag: float,
) -> list[Face3D]:
    """Keep only the outermost faces in a cluster (exterior wall faces).

    For each direction cluster, interior wall faces sit at a different depth
    (distance along the facade direction) than the exterior walls.  We keep
    faces whose depth is within a tolerance of the maximum depth (the
    outermost/exterior position).

    Tolerance is 15% of the total depth span of the cluster, or 0.6 m,
    whichever is larger.  This accommodates wall thickness, window recesses,
    and minor offsets while filtering out interior partitions that are
    well behind the exterior wall.
    """
    if not cluster_faces:
        return cluster_faces

    depths = [dot(_face_center(f), direction) for f in cluster_faces]
    max_depth = max(depths)
    min_depth = min(depths)
    span = max_depth - min_depth

    if span < 0.01:
        # All faces at roughly the same depth — keep all.
        return cluster_faces

    # Tolerance: 15% of the cluster's depth span, min 0.6 m.
    tol = max(span * 0.15, 0.6)

    return [
        f for f, d in zip(cluster_faces, depths)
        if d >= max_depth - tol
    ]


def _extract_facades_with_axis(
    faces: list[Face3D], up_axis: Literal["Y", "Z"]
) -> list[Facade]:
    """Extract facades assuming a specific up axis."""

    # 1. Filter vertical faces.
    vertical_faces: list[Face3D] = []
    for face in faces:
        up_comp = _get_up_component(face.normal, up_axis)
        if abs(up_comp) <= VERTICAL_EPSILON:
            vertical_faces.append(face)

    if not vertical_faces:
        return []

    model_diag = _compute_model_diagonal(faces)

    # 2. Cluster by horizontal direction.
    clusters = _cluster_by_direction(vertical_faces, up_axis)

    # 3. Build one Facade per cluster.
    facades: list[Facade] = []

    for direction, cluster_faces in clusters:
        # 3a. Filter to keep only exterior (outermost) faces per cluster.
        cluster_faces = _filter_exterior_faces(cluster_faces, direction, model_diag)
        u_axis, v_axis = _compute_facade_axes(direction, up_axis)

        # Project all face vertices to the facade's 2D space.
        # Use the model origin as projection reference.
        origin = Vec3(0.0, 0.0, 0.0)

        polygons: list[Loop2D] = []
        all_u: list[float] = []
        all_v: list[float] = []

        for face in cluster_faces:
            pts_2d: list[Vec2] = []
            for v in face.vertices:
                u = dot(v, u_axis)
                vv = dot(v, v_axis)
                pts_2d.append(Vec2(u, vv))
                all_u.append(u)
                all_v.append(vv)
            if len(pts_2d) >= 3:
                polygons.append(Loop2D(vertices=pts_2d, panel_id=face.panel_id))

        if not polygons or not all_u:
            continue

        # Compute bounding box.
        min_u = min(all_u)
        max_u = max(all_u)
        min_v = min(all_v)
        max_v = max(all_v)

        width = max_u - min_u
        height = max_v - min_v

        if width < 0.01 or height < 0.01:
            continue

        # Normalize: shift so that (0,0) = bottom-left.
        normalized_polygons: list[Loop2D] = []
        for poly in polygons:
            normalized_polygons.append(
                Loop2D(
                    vertices=[Vec2(p.x - min_u, p.y - min_v) for p in poly.vertices],
                    panel_id=poly.panel_id,
                )
            )

        label = _direction_label(direction, up_axis)
        # Ensure unique labels.
        existing_labels = {f.label for f in facades}
        if label in existing_labels:
            n = 2
            while f"{label} {n}" in existing_labels:
                n += 1
            label = f"{label} {n}"

        facades.append(Facade(
            label=label,
            direction=direction,
            polygons=normalized_polygons,
            width=width,
            height=height,
        ))

    # Sort facades by label for consistent ordering.
    order = {"Norte": 0, "Este": 1, "Sur": 2, "Oeste": 3}
    facades.sort(key=lambda f: order.get(f.label.split()[-1], 99))

    return facades


def extract_facades(
    faces: list[Face3D], up_axis: Literal["Y", "Z"] | None = None
) -> list[Facade]:
    """Extract facades from faces.

    If *up_axis* is provided it is used directly; otherwise both Y and Z are
    tried and the one producing more geometry wins.
    """
    result, _ = extract_facades_with_detected_axis(faces, up_axis)
    return result


def extract_facades_with_detected_axis(
    faces: list[Face3D], up_axis: Literal["Y", "Z"] | None = None
) -> tuple[list[Facade], Literal["Y", "Z"]]:
    """Like extract_facades but also returns the up axis that was used.

    The pipeline calls this so it can re-use the same axis for floor plans
    and decomposition, ensuring all views are consistent.
    """
    if not faces:
        return [], "Z"

    if up_axis is not None:
        return _extract_facades_with_axis(faces, up_axis), up_axis

    # Try both up-axis conventions and pick the best.
    facades_z = _extract_facades_with_axis(faces, "Z")
    facades_y = _extract_facades_with_axis(faces, "Y")

    # Compare by total facade area (width × height).  Using area instead of
    # polygon count avoids false positives when slab faces get misclassified
    # as vertical under the wrong axis — their projected "facades" are small.
    area_z = sum(f.width * f.height for f in facades_z)
    area_y = sum(f.width * f.height for f in facades_y)

    if area_y > area_z:
        return facades_y, "Y"
    return facades_z, "Z"
